src/Simulation/Common/IQuantumProcessor.cs

```diff
@@ -391,6 +391,41 @@ public interface IQuantumProcessor
+        /// A variety of parameter types are provided for flexibility.
+        /// </remarks>
+        /// <param name="gateId">The id of the custom gate. It is up to the custom gate implementor to define the meaning.</param>
+        /// <param name="qubits">Array of qubits to apply the gate to. Their meaning is specific to the custom gate.</param>
+        /// <param name="paulis">Array of single-qubit Pauli values. Their meaning is specific to the custom gate.</param>
+        /// <param name="longs">Array of long parameters. Their meaning is specific to the custom gate.</param>
+        /// <param name="doubles">Array of double parameters. Their meaning is specific to the custom gate.</param>
+        /// <param name="bools">Array of bool parameters. Their meaning is specific to the custom gate.</param>
+        /// <param name="results">Array of Result parameters. Their meaning is specific to the custom gate.</param>
+        /// <param name="strings">Array of string parameters. Their meaning is specific to the custom gate.</param>
+        void Gate(long gateId, IQArray<Qubit> qubits, IQArray<Pauli> paulis, IQArray<long> longs, IQArray<double> doubles, IQArray<bool> bools, IQArray<Result> results, IQArray<string> strings);
+
+        /// <summary>
+        /// Used as extensibility mechanism to add custom gates to IQuantumProcessor. Implements controlled version.
+        /// </summary>
+        /// <remarks>
+        /// The custom gate may use only some of the parameters. Depending on the nature of the gate, all of them may be optional. 
+        /// A variety of parameter types are provided for flexibility.
+        /// </remarks>
+        /// <param name="gateId">The id of the custom gate. It is up to the custom gate implementor to define the meaning.</param>
+        /// <param name="controls">Array of control qubits for this operation.</param>
+        /// <param name="qubits">Array of qubits to apply the gate to. Their meaning is specific to the custom gate.</param>
+        /// <param name="paulis">Array of single-qubit Pauli values. Their meaning is specific to the custom gate.</param>
+        /// <param name="longs">Array of long parameters. Their meaning is specific to the custom gate.</param>
+        /// <param name="doubles">Array of double parameters. Their meaning is specific to the custom gate.</param>
+        /// <param name="bools">Array of bool parameters. Their meaning is specific to the custom gate.</param>
+        /// <param name="results">Array of Result parameters. Their meaning is specific to the custom gate.</param>
+        /// <param name="strings">Array of string parameters. Their meaning is specific to the custom gate.</param>
+        void ControlledGate(long gateId, IQArray<Qubit> controls, IQArray<Qubit> qubits, IQArray<Pauli> paulis, IQArray<long> longs, IQArray<double> doubles, IQArray<bool> bools, IQArray<Result> results, IQArray<string> strings);
```

Comment:
Why are there methods for applying the body and controlled body, but not the adjoint and controlled adjoint? Also, the same comment applies here about providing a default implementation to avoid breaking changes to the interface.

Reply:
Valid concern! I deliberated on it for quite some time and decided for controlled but against adjoint. We do not provide adjoint for all gates, so this is not expected. There are two ways to implement adjoint of such custom gate: One is to add Adjoint as another interface function (actually two: Adjoint and AdjointControled). Another way is to treat it as a gate with a different id. The second one looked appropriate and easier. The drawback of the first one is addition of two functions to the API, which could be avoided. Controlled is provided separately because it adds another parameter - an array of control qubits. 
So, I could go either way, but I think current one is better.

---
In reply to: [410524847](https://github.com/microsoft/qsharp-runtime/pull/172#discussion_r410524847) [](ancestors = 410524847)